hugr-model/tests/fixtures/model-constraints.edn

```diff
@@ -0,0 +1,15 @@
+(hugr 0)
+
+(declare-func array.replicate
+  (forall ?t type)
+  (forall ?n nat)
+  (where (copy ?t))
+  (where (discard ?t))
```

Comment:
this looks like it could get quite verbose. Will we need to specify things that are `copy` but not `discard`?  Even if we do, I feel that ought to be the more verbose specification. Ideally I guess I want to write:
1. Nothing for things that are copyable and discardable
2. "linear" or "not copy" for things that are such
3. "affine" when I can't copy but can discard.

Is this just a handy syntax change or are we running in to subtyping problems again?

Reply:
Yeah, I feel that the "sales blurb" in the description that "The translation in this PR requires the two to be used together since TypeBound does not have the facility to express one without the other. Once we move to constraints in hugr-core, that restriction would be lifted." misses the point. We *could* have had separate copy/discard/affine/etc. in TypeBound, but [we discussed this](https://github.com/CQCL/hugr/discussions/1298) and did not decide that we wanted it. So I see no reason to do so now.

Indeed, if we change our minds are decide we want to add this in the future, there is still an easy mechanism (right?) - we just add two new constraints (say `copy` and `discard`) alongside the old `copyable` and then define that `copyable` holds if both `copy` and `discard` hold (and that each of the latter/two hold if `copyable` does), right? (so ok we need a cyclicity check in the constraint solver.) The problem would just be better naming, we can't really have both `copy` and `copyable`....

So (the more I think about it, the more) I think we should stick with one combined constraint rather than two for now.

Reply:
`nonlinear` ..?

Reply:
I'm fine with `nonlinear`. @acl-cqc ?

Reply:
The two constraints are joined into a `nonlinear` constraint now.